hivemq-edge/src/main/java/com/hivemq/persistence/topicfilter/TopicFilterPersistenceReaderWriter.java

```diff
@@ -0,0 +1,124 @@
+import com.hivemq.configuration.info.SystemInformation;
+import com.hivemq.exceptions.UnrecoverableException;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.persistence.topicfilter.xml.TopicFilterPersistenceEntity;
+import com.hivemq.persistence.topicfilter.xml.TopicFilterXmlEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Singleton
+public class TopicFilterPersistenceReaderWriter {
+
+    private static final Logger log = LoggerFactory.getLogger(TopicFilterPersistenceReaderWriter.class);
+    public static final String PERSISTENCE_FILE_NAME = "topic-filters.xml";
+    private final @NotNull File persistenceFile;
+    private final @NotNull XmlMapper xmlMapper = new XmlMapper();
+
+    @Inject
+    public TopicFilterPersistenceReaderWriter(
+            final @NotNull SystemInformation systemInformation) {
+        this.persistenceFile = new File(systemInformation.getConfigFolder(), PERSISTENCE_FILE_NAME);
+        xmlMapper.setAnnotationIntrospector(AnnotationIntrospector.pair(new JacksonXmlAnnotationIntrospector(),
+                new JaxbAnnotationIntrospector(TypeFactory.defaultInstance())));
+        xmlMapper.enable(SerializationFeature.INDENT_OUTPUT);
+        xmlMapper.configure(ToXmlGenerator.Feature.WRITE_XML_DECLARATION, true);
+    }
+
+    public @NotNull List<TopicFilter> readPersistence() {
+        final TopicFilterPersistenceEntity persistenceEntity = readPersistenceXml();
+        return convertToTopicFilters(persistenceEntity);
+    }
+
+    public synchronized void writePersistence(final @NotNull Collection<TopicFilter> tags) {
+        if (persistenceFile.exists() && !persistenceFile.canWrite()) {
+            log.error("Unable to write to persistence file {}, because it is not writable.", persistenceFile);
+            throw new UnrecoverableException(false);
+        }
+        log.debug("Writing persistence file {}", persistenceFile);
+
+        try {
+            final TopicFilterPersistenceEntity persistenceEntity = convertToEntity(tags);
+            final String xml = xmlMapper.writeValueAsString(persistenceEntity);
+            Files.writeString(persistenceFile.toPath(), xml);
+        } catch (final JsonProcessingException e) {
+            log.error(
+                    "Error while trying to persist the topic filters on disc. Exception happened during serialization of topic filters:",
+                    e);
+            throw new RuntimeException(e);
+        } catch (final IOException e) {
+            log.error("Error while trying to persist the topic filters on disc. Exception happened during writing of " +
+                    PERSISTENCE_FILE_NAME +
+                    ":", e);
+            throw new RuntimeException(e);
```

Comment:
Why not use UnrecoverableException here?

Reply:
Wont do something outside of startup